documentcloud/entities/views.py

```diff
@@ -1,56 +1,122 @@
 # Django
-from rest_framework import permissions, viewsets
-from rest_framework.decorators import api_view
-from rest_framework.response import Response
-from rest_framework.reverse import reverse
+from django.utils.decorators import method_decorator
+from rest_framework import exceptions, serializers, viewsets
+from rest_framework.generics import get_object_or_404
+from rest_framework.permissions import SAFE_METHODS
+
+# Standard Library
+from functools import lru_cache
+
+# Third Party
+from django_filters import rest_framework as django_filters
 
 # DocumentCloud
-from documentcloud.entities.models import Entity, EntityOccurrence2
-from documentcloud.entities.permissions import IsOwnerOrReadOnly
+from documentcloud.documents.decorators import conditional_cache_control
+from documentcloud.documents.models import Document
+from documentcloud.drf_bulk.views import BulkCreateModelMixin
+from documentcloud.entities.choices import EntityAccess
+from documentcloud.entities.models import Entity, EntityOccurrence
 from documentcloud.entities.serializers import (
-    EntityOccurrence2Serializer,
+    EntityOccurrenceSerializer,
     EntitySerializer,
 )
 
 
-class EntityViewSet(viewsets.ModelViewSet):
+class EntityViewSet(BulkCreateModelMixin, viewsets.ModelViewSet):
     serializer_class = EntitySerializer
-    permission_classes = [permissions.IsAuthenticatedOrReadOnly]
+    queryset = Entity.objects.none()
 
     def get_queryset(self):
-        queryset = Entity.objects.all()
-        wikidata_id = self.request.query_params.get("wikidata_id")
-        name = self.request.query_params.get("name")
+        return Entity.objects.get_viewable(self.request.user)
+
+    def perform_create(self, serializer):
+        if serializer.validated_data.get("access") == EntityAccess.private:
+            # set the user on prviate entities
+            entity = serializer.save(user=self.request.user)
+        else:
+            # lookup wikidata on public entities
+            entity = serializer.save()
+            try:
+                entity.lookup_wikidata()
+                entity.save()
+            except ValueError:
+                raise serializers.ValidationError("Invalid `wikidata_id`")
 
-        if wikidata_id:
-            queryset = queryset.filter(wikidata_id=wikidata_id)
-        if name:
-            queryset = queryset.filter(name=name)
+    def bulk_perform_create(self, serializer):
+        first = serializer.validated_data[0]
+        if not all(
+            d.get("access") == first.get("access") for d in serializer.validated_data
+        ):
+            raise serializers.ValidationError(
+                "Bulk entity creation must all have same `access`"
+            )
+        if first.get("access") == EntityAccess.private:
```

Comment:
I believe I was trying to determine it before saving the model, hence the logic needed to be outside.  But then I ended up just saving the model anyway before doing the wikidata lookup, as it started getting too complicated - so there is an extra DB call there.  And the bulk version should probably be optimized into making less DB calls.  This should be cleaned up.  The bulk tends to be a pain - you either need to accept a performance hit and do each one individually, or re-write some logic so it can be applied in bulk more efficiently.

Reply:
Oh, OK, so someday, the bulk create will someday be different (and may fetch from wikidata in batch?). Handy to know.